plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.ts

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2021 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { LocationSpec } from '@backstage/catalog-model';
+import { Config } from '@backstage/config';
+import { graphql } from '@octokit/graphql';
+import { Logger } from 'winston';
+import {
+  getOrganizationRepositories,
+  ProviderConfig,
+  readGithubConfig,
+} from './github';
+import * as results from './results';
+import { CatalogProcessor, CatalogProcessorEmit } from './types';
+
+/**
+ * Extracts teams and users out of a GitHub org.
+ */
+export class GithubDiscoveryProcessor implements CatalogProcessor {
+  private readonly providers: ProviderConfig[];
+  private readonly logger: Logger;
+
+  static fromConfig(config: Config, options: { logger: Logger }) {
+    return new GithubDiscoveryProcessor({
+      ...options,
+      providers: readGithubConfig(config),
```

Comment:
This probably isn't the config you want. It's the one that the Org plugins use, but you actually want to use the `integration` config - which is what pretty much everybody uses. This is how you get auth in the right way at the moment:

```ts
const configs = readGitHubIntegrationConfigs(
  config.getOptionalConfigArray('integrations.github') ?? [],
);
const providersByHost = new Map(configs.map(c => [c.host, GithubCredentialsProvider.create(c)]));
```

now you can find a credential provider by hostname (NOT by url prefix - just the hostname - use something like `new URL(location.target).hostname` or something), and do `credentialProvider.getCredentials({ url: location.target })` after that. You can therefore remove the code below that looks at configs etc, all you need should be this map.

I realize this was a lot of noise to do something simple. I'll look into if we can wrap the whole map-of-providers thing wrapped up in one easy to use package. But the benefit of doing the above, is that you'll get automatic support for github apps authentication and future features.

Reply:
I wonder though how this should work with GitHub Enterprise and GraphQL endpoints. From the docs it seems the API endpoint should be something like `https://api.github.com/graphql`, but the API base URL in the examples shown in the `app-config.yaml` is `https://ghe.example.net/api/v3` or `https://ghe.example.net/raw`. Should I convert it somehow? Should it be part of the Credentials Provider? Maybe it's a good opportunity to create a wrapper?

Reply:
Well for GHE it is meant to use the same apiBaseUrl, and that's something that already works.

I don't know of a specific graphql endpoint.

Reply:
NM, I see that Octokit handles this well:
https://github.com/octokit/graphql.js/blob/master/src/graphql.ts#L53